ddl/lightning/backend_mgr.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/pingcap/tidb/br/pkg/lightning/errormanager"
+	"github.com/pingcap/tidb/br/pkg/lightning/glue"
+	"github.com/pingcap/tidb/br/pkg/lightning/log"
+	"github.com/pingcap/tidb/parser"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/util/generic"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+type backendCtxManager struct {
+	generic.SyncMap[int64, *BackendContext]
+	memRoot  MemRoot
+	diskRoot DiskRoot
+}
+
+func (m *backendCtxManager) init(memRoot MemRoot, diskRoot DiskRoot) {
+	m.SyncMap = generic.NewSyncMap[int64, *BackendContext](10)
+	m.memRoot = memRoot
+	m.diskRoot = diskRoot
+}
+
+// Register creates a new backend and registers it to the backend context.
+func (m *backendCtxManager) Register(ctx context.Context, unique bool, jobID int64, _ mysql.SQLMode) (*BackendContext, error) {
+	bc, exist := m.Load(jobID)
+	if !exist {
+		m.memRoot.RefreshConsumption()
+		ok := m.memRoot.CheckConsume(StructSizeBackendCtx)
+		if !ok {
+			return nil, genBackendAllocMemFailedErr(m.memRoot, jobID)
+		}
+		cfg, err := generateLightningConfig(m.memRoot, jobID, unique)
+		if err != nil {
+			logutil.BgLogger().Warn(LitWarnConfigError, zap.Int64("job ID", jobID), zap.Error(err))
+			return nil, err
+		}
+		bd, err := createLocalBackend(ctx, cfg, glueLit{})
+		if err != nil {
+			logutil.BgLogger().Error(LitErrCreateBackendFail, zap.Int64("job ID", jobID), zap.Error(err))
+			return nil, err
+		}
+
+		bcCtx := newBackendContext(ctx, jobID, &bd, cfg, defaultImportantVariables, m.memRoot, m.diskRoot)
+		m.Store(jobID, bcCtx)
+
+		m.memRoot.Consume(StructSizeBackendCtx)
+		logutil.BgLogger().Info(LitInfoCreateBackend, zap.Int64("job ID", jobID),
+			zap.Int64("current memory usage", m.memRoot.CurrentUsage()),
+			zap.Int64("max memory quota", m.memRoot.MaxMemoryQuota()),
+			zap.Bool("is unique index", unique))
+		return bcCtx, nil
+	}
+	return bc, nil
+}
+
+func createLocalBackend(ctx context.Context, cfg *config.Config, glue glue.Glue) (backend.Backend, error) {
+	tls, err := cfg.ToTLS()
+	if err != nil {
+		logutil.BgLogger().Error(LitErrCreateBackendFail, zap.Error(err))
+		return backend.Backend{}, err
+	}
+
+	errorMgr := errormanager.New(nil, cfg, log.Logger{Logger: logutil.BgLogger()})
+	return local.NewLocalBackend(ctx, tls, cfg, glue, int(LitRLimit), errorMgr)
+}
+
+func newBackendContext(ctx context.Context, jobID int64, be *backend.Backend,
+	cfg *config.Config, vars map[string]string, memRoot MemRoot, diskRoot DiskRoot) *BackendContext {
+	bc := &BackendContext{
+		jobID:    jobID,
+		backend:  be,
+		ctx:      ctx,
+		cfg:      cfg,
+		sysVars:  vars,
+		diskRoot: diskRoot,
+	}
+	bc.EngMgr.init(memRoot, diskRoot)
+	return bc
+}
+
+// Unregister removes a backend context from the backend context manager.
+func (m *backendCtxManager) Unregister(jobID int64) {
+	bc, exist := m.Load(jobID)
+	if !exist {
+		return
+	}
+	bc.EngMgr.UnregisterAll(jobID)
```

Comment:
I don't see where `EngMgr.Register` is called. Has it not been called yet or am I reading it wrong?

Reply:
The `engineInfo` is initialized in `writePhysicalTableRecord`. It will be used in the next PR:

https://github.com/pingcap/tidb/pull/36767/files#diff-1bbc7833321782b3a22a842b6237974b6a60cb07b49b246b85552ab5e915ed6eR757